docs/tsm/TSM_FORMAT.md

```diff
@@ -0,0 +1,161 @@
+│4 bytes│ N*4 bytes  │2 bytes│N bytes │4 bytes │N bytes│    │
+└───────┴────────────┴───────┴────────┴────────┴───────┴────┘
+
+A `Values` block contains a sorted set of values for a single tag.  It has a `Hash Index` into the `Sorted Value` section.
+
+The `Sorted Value/Block List` section contains a 2 byte length followed by the text value bytes.  This is followed by a 4 byte length and a compressed bitmap containing Series offsets.  They always point to the offset in the file where the encoded Series exists.
+
+An alternative option is to store uncompressed 4 byte pointers instead of a compressed bitmap.
+
+┌───────────────────────────────────────────────────┐
+│                       Tags                        │
+├────────────────────┬─────────────────────────┬────┤
+│     Hash Index     │       Sorted Tags       │    │
+├───────┬────────────┼───────┬────────┬────────┤... │
+│  Len  │ Hash Array │  Len  │  Tag   │Val Ofs │    │
+│4 bytes│ N*4 bytes  │2 bytes│N bytes │4 bytes │    │
+└───────┴────────────┴───────┴────────┴────────┴────┘
+
+The `Tags` section is similar to the `Values` section in that it first has a
+`Hash Index` containing offsets into the tags block.  The `Sorted Tags` section contains a sorted list of tag length and tag bytes followed by a 4 byte offset pointing to the start of the `Values` block for that tag.
+
+┌────────────────────────────────────────────────────────────────────┐
+│                            Measurement                             │
+├────────────────────┬──────────────────────────────────────────┬────┤
+│     Hash Index     │           Sorted Measurements            │    │
+├───────┬────────────┼───────┬────────┬────────┬────────┬───────┤... │
+│  Len  │ Hash Array │  Len  │  Meas  │Tag Ofs │  Len   │Bitmap │    │
+│4 bytes│ N*4 bytes  │2 bytes│N bytes │4 bytes │4 bytes │4 bytes│    │
+└───────┴────────────┴───────┴────────┴────────┴────────┴───────┴────┘
+
+The `Measurement` block continas a `Hash Index` into the `Sorted Measurements` section.  The `Sorted Measurements` sections is an ordered list of measurement names followed by a 4 byte offset pointing to the start of the `Tags` section for that measurement.  It also contains a compressed bitmap of offsets into the `Series` that maps all series in the measurment for the cases where a measurement has no tags.
+
+## Tombstones
+
+TODO: Fill this out more (Use FLAG idea already proposed)
+
+## Index Construction
+
+TODO: Fill this out more (Use compactions/WAL idea already proposed)
+
+## Hash Indexes
+
+The `Measurement`, `Tags` and `Values` sections all contain hash indexes hashing the key to an offset in the sorted data section.  The hashing algorithm used would be _Robin Hood Hashing_ [1] which minimizes worst case search time and is also more CPU/Disk cache friendly.  This allows for O(1) lookups when using exact filtering (very common) and more closely matches the current in-memory structure.
```

Comment:
If we're using 4 bytes to store an offset then we are limiting the maximum size of a TSI file to somewhere ~4GB. It's hard to say where because we have to store the `Series Dictionary` as well as the other offset-containing blocks. 

Is that OK? This comes back to the question of addressable space on 32bit systems. If we're supporting 32bit systems then I guess it's fine.